lib/filterx/filterx-metrics-labels.c

```diff
@@ -0,0 +1,361 @@
+  scratch_buffers_reclaim_marked(marker);
+
+  return !!self->value.str;
+}
+
+static FilterXMetricsLabel *
+_label_new(FilterXExpr *name, FilterXExpr *value)
+{
+  FilterXMetricsLabel *self = g_new0(FilterXMetricsLabel, 1);
+
+  self->name = _init_label_name(name);
+  if (!self->name)
+    goto error;
+
+  if (!_init_label_value(self, value))
+    goto error;
+
+  return self;
+
+error:
+  _label_free(self);
+  return NULL;
+}
+
+
+struct _FilterXMetricsLabels
+{
+  FilterXExpr *expr;
+  GPtrArray *literal_labels;
+  gboolean is_const;
+};
+
+static void
+_check_label_is_const(gpointer data, gpointer user_data)
+{
+  FilterXMetricsLabel *label = (FilterXMetricsLabel *) data;
+  gboolean *is_const = (gboolean *) user_data;
+
+  if (!(*is_const))
+    return;
+
+  if (!_label_is_const(label))
+    *is_const = FALSE;
+}
+
+gboolean
+filterx_metrics_labels_is_const(FilterXMetricsLabels *self)
+{
+  return self->is_const;
+}
+
+static gboolean
+_format_dict_elem_to_tls_cache(FilterXObject *key, FilterXObject *value, gpointer user_data)
+{
+  const gchar *name_str = _format_str_obj(key);
+  if (!name_str)
+    {
+      filterx_eval_push_error_info("failed to format label name, name must be string", NULL,
+                                   g_strdup_printf("got %s instead", key->type->name), TRUE);
+      return FALSE;
+    }
+
+  const gchar *value_str = _format_str_obj(value);
+  if (!value_str)
+    {
+      filterx_eval_push_error_info("failed to format label value, object does not support string casting", NULL,
+                                   "", FALSE);
+      return FALSE;
+    }
+
+  StatsClusterLabel *label = metrics_cache_alloc_label(metrics_tls_cache());
+  label->name = name_str;
+  label->value = value_str;
+
+  return TRUE;
+}
+
+static gboolean
+_format_expr_to_tls_cache(FilterXExpr *expr)
+{
+  FilterXObject *obj = filterx_expr_eval_typed(expr);
+  if (!obj)
+    return FALSE;
+
+  gboolean success = FALSE;
+
+  if (!filterx_object_is_type(obj, &FILTERX_TYPE_NAME(dict)))
+    {
+      filterx_eval_push_error_info("failed to format metrics labels, labels must be a dict", expr,
+                                   g_strdup_printf("got %s instead", obj->type->name), TRUE);
+      goto exit;
+    }
+
+  success = filterx_dict_iter(obj, _format_dict_elem_to_tls_cache, NULL);
+  if (!success)
+    goto exit;
+
+  metrics_cache_sort_labels(metrics_tls_cache());
+
+exit:
+  filterx_object_unref(obj);
+  return success;
+}
+
+static void
+_format_label_to_tls_cache(gpointer data, gpointer user_data)
+{
+  FilterXMetricsLabel *label = (FilterXMetricsLabel *) data;
+  gboolean *success = (gboolean *) user_data;
+
+  if (!(*success))
+    return;
+
+  *success = _label_format(label, metrics_cache_alloc_label(metrics_tls_cache()));
+}
+
+gboolean
+filterx_metrics_labels_format(FilterXMetricsLabels *self, StatsClusterLabel **labels, gsize *len)
+{
+  metrics_cache_reset_labels(metrics_tls_cache());
```

Comment:
is there a point in saving the metrics_tls_cache() value and not do it 3x?

Reply:
Thanks, fixed in a new commit.